File: app/src/main/java/de/westnordost/streetcomplete/data/QuestTypeRegistry.java
package de.westnordost.streetcomplete.data;

import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/** Every osm quest needs to be registered here.
 *
 *  Could theoretically be done with Reflection, but that doesn't really work on Android
 */

public class QuestTypeRegistry
{
	private final List<QuestType> types;
	private final Map<String, QuestType> typeMap;

	public QuestTypeRegistry(List<QuestType> types)
	{
		this.types = types;

		typeMap = new HashMap<>();
		for (QuestType questType : types)
		{
			String questTypeName = questType.getClass().getSimpleName();
			if(typeMap.containsKey(questTypeName))
			{
				throw new RuntimeException("A quest type's name must be unique! \"" +
						questTypeName + "\" is defined twice!");
			}
			typeMap.put(questTypeName, questType);
		}
	}

	public QuestType getByName(String typeName)
	{
		return typeMap.get(typeName);
	}

	public List<QuestType> getAll()
	{
		return Collections.unmodifiableList(types);
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/QuestTypeRegistry.kt
package de.westnordost.streetcomplete.data

/** Every osm quest needs to be registered here.
 *
 * Could theoretically be done with Reflection, but that doesn't really work on Android
 */

class QuestTypeRegistry(val all: List<QuestType<*>>) {
    private val typeMap: Map<String, QuestType<*>>

    init {
        val map = mutableMapOf<String, QuestType<*>>()
        for (questType in all) {
            val questTypeName = questType.javaClass.simpleName
            require(!map.containsKey(questTypeName)) {
                "A quest type's name must be unique! \"$questTypeName\" is defined twice!"
            }
            map[questTypeName] = questType
        }
        typeMap = map
    }

    fun getByName(typeName: String): QuestType<*>? {
        return typeMap[typeName]
    }
}
